src/Core/Grid/Query/ProductCombinationQueryBuilder.php

```diff
@@ -0,0 +1,164 @@
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Grid\Query;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+
+final class ProductCombinationQueryBuilder extends AbstractDoctrineQueryBuilder
+{
+    /**
+     * @var DoctrineSearchCriteriaApplicatorInterface
+     */
+    private $searchCriteriaApplicator;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+     */
+    public function __construct(
+        Connection $connection,
+        string $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+    ) {
+        parent::__construct($connection, $dbPrefix);
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getCombinationsQueryBuilder($searchCriteria)->select('pa.*');
+
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        return $this->getCombinationsQueryBuilder($searchCriteria)
+            ->select('COUNT(pa.id_product_attribute)')
+        ;
+    }
+
+    /**
+     * @param SearchCriteriaInterface $searchCriteria
+     *
+     * @return QueryBuilder
+     */
+    private function getCombinationsQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $filters = $searchCriteria->getFilters();
+        if (!isset($filters['product_id'])) {
+            //@todo: better exception? or require filter in another layer - like builder for filters? (categories has same issue)
+            throw new \RuntimeException('Product id is required for combinations grid');
```

Comment:
I think after @jolelievre implemented CombinationFilters and it requires productId in constructor we don't need this check anymore

Reply:
Indeed this must be removed

Reply:
Ideally the builder should use `ProductCombinationFilters` as a parameter instead of the interface And we could add a getter function `getProductId` in the filters class

But I don't know if it could cause some error for type mismatching with the interface prototype, this should be tested